t/08_new_api.t

```diff
@@ -0,0 +1,305 @@
+    is $res->{EMAIL_KEY()},   FIRST_EMAIL, 'Email is retrieved successfully';
+    is $res->{REFRESH_KEY()}, REFRESH_SET, 'Refresh is retrieved successfully';
+    exists $res->{'_global_rev'}, 'Batch get returns global revision';
+};
+
+subtest 'Gets/sets of illegal keys' => sub {
+    subtest 'Attempt to set static key' => sub {
+        my $app_config = _new_app_config();
+        throws_ok {
+            $app_config->set({ADMINS_KEY() => ADMINS_SET});
+        }
+        qr/Cannot set with key/;
+    };
+
+    subtest 'Attempt to set non-existent key' => sub {
+        my $app_config = _new_app_config();
+        throws_ok {
+            $app_config->set({NON_EXT_KEY() => ADMINS_SET});
+        }
+        qr/Cannot set with key/;
+    };
+
+    subtest 'Attempt to get non-existent key' => sub {
+        my $app_config = _new_app_config();
+        throws_ok {
+            $app_config->get(NON_EXT_KEY);
+        }
+        qr/Cannot get with key/;
+    };
+};
+
+subtest 'History chronicling' => sub {
+    my $app_config = _new_app_config();
+    my $module     = Test::MockModule->new('Data::Chronicle::Reader');
+
+    subtest 'Add history of values' => sub {
+        # Sleeps ensure the chronicle records them at different times
+        ok $app_config->set({EMAIL_KEY() => FIRST_EMAIL}), 'Set 1st value succeeds';
+        sleep 1;
+        ok $app_config->set({EMAIL_KEY() => SECOND_EMAIL}), 'Set 2nd value succeeds';
+        sleep 1;
+        ok $app_config->set({EMAIL_KEY() => THIRD_EMAIL}), 'Set 3rd value succeeds';
+        sleep 1;
+    };
+
+    subtest 'Get history' => sub {
+        is($app_config->get_history(EMAIL_KEY, 0, 1), THIRD_EMAIL,  'History retrieved successfully');
+        is($app_config->get_history(EMAIL_KEY, 1, 1), SECOND_EMAIL, 'History retrieved successfully');
+        is($app_config->get_history(EMAIL_KEY, 2, 1), FIRST_EMAIL,  'History retrieved successfully');
+    };
+
+    subtest 'Ensure get_history is cached (i.e. get_history should not be called)' => sub {
+        $module->mock('get_history', sub { ok(0, 'get_history should not be called here') });
+        is($app_config->get_history(EMAIL_KEY, 0), THIRD_EMAIL,  'Email retrieved via cache');
+        is($app_config->get_history(EMAIL_KEY, 1), SECOND_EMAIL, 'Email retrieved via cache');
+        is($app_config->get_history(EMAIL_KEY, 2), FIRST_EMAIL,  'Email retrieved via cache');
+        $module->unmock('get_history');
+    };
+
+    subtest 'Ensure cache goes stale when new is set' => sub {
+        is($app_config->get_history(EMAIL_KEY, 1, 1), SECOND_EMAIL, 'Previous email is correct');
+        ok $app_config->set({EMAIL_KEY() => FIRST_EMAIL}), 'Set email succeeds';
+        is($app_config->get_history(EMAIL_KEY, 1), THIRD_EMAIL, 'Correct previous email is returned');
+    };
+
+    subtest 'Check caching can be disabled' => sub {
+        plan tests => 5;    # Ensures the ok checks inside the mocked sub are run
```

Comment:
FYI another way to avoid that would be some logic like

    my $get_history_called;
    $module->mock(get_history => sub { $get_history_called++; return {data => SECOND_EMAIL}});
    ...
    ok $get_history_called, 'get_history was called';

Keeps the test more flexible in light of further additions